url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/grammaticalbehaviors/GEBT_Mario/Map.java

```diff
@@ -0,0 +1,1047 @@
+            {
+                m_enemyCollection.put(m_currentXWritting, new Vector<Enemy>());
+            }
+            m_enemyCollection.get(m_currentXWritting).add(new Enemy(m_currentXWritting, a_y, a_data));
+        }
+    }
+
+    public boolean checkForEnemyType(int a_xSrc, int a_xDest, int a_ySrc, int a_yDest, int a_type)
+    {
+        int min_x = Math.min(a_xSrc, a_xDest);
+        int min_y = Math.min(a_ySrc, a_yDest);
+        int max_x = Math.max(a_xSrc, a_xDest);
+        int max_y = Math.max(a_ySrc, a_yDest);
+
+        if(min_x < 0) min_x = 0;
+        if(max_y >= 15) max_y = 14;
+        if(min_y < 0) min_y = 0;
+
+
+        for(int x = min_x; x <= max_x; ++x)
+        {
+            for(int y = min_y; y <= max_y; ++y)
+            {
+                if(m_enemyCollection.containsKey(x))
+                {
+                    Vector<Enemy> enemies = m_enemyCollection.get(x);
+                    for(int i = 0; i < enemies.size(); ++i)
+                    {
+                        Enemy en = enemies.elementAt(i);
+                        if(en.m_type == a_type && en.m_y == y)
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+        
+    }
+
+
+    public void updateLastCell()
+    {
+        //Update last cell written
+        if(m_currentXWritting > m_lastCell)
+            m_lastCell = m_currentXWritting;
+    }
+
+    public byte read(int a_x, int a_y)
+    {
+        //Check boundaries
+        if(a_y < 0 || a_y >= LEVEL_HEIGHT)
+            return -1;
+
+        //Get the block for this position
+        int blockIndex = (int)(a_x / MAP_BLOCK_LENGTH);
+        int xInBlock = (int)(a_x % MAP_BLOCK_LENGTH);
+        if(blockIndex < m_levelMap.size())
+        {
+            return getBlock(blockIndex).read(xInBlock,a_y);
+        }
+        
+        return -1;
+    }
+
+    public void dumpToFile(String filename)
+    {
+        try
+        {
+            FileWriter fw = new FileWriter(filename);
+            BufferedWriter bw = new BufferedWriter(fw);
+
+            for(int h = 14; h >= 0; --h)
+            {
+                for(int w = 0; w < m_lastCell; ++w)
+                {
+                    //Get the block for this position
+                    int blockIndex = (int)(w / MAP_BLOCK_LENGTH);
+                    int xInBlock = (int)(w % MAP_BLOCK_LENGTH);
+                    getBlock(blockIndex).dumpToFile(bw,xInBlock,h);
+                }
+
+                //End of this line
+                bw.newLine();
+            }
+
+            //And that's all!
+            bw.close();
+
+        }catch(Exception e)
+        {
+            e.printStackTrace();
+        }
+    }
+
+    public void dumpToFileProcessing(String filename)
+    {
+        try
+        {
+            FileWriter fw = new FileWriter(filename);
+            BufferedWriter bw = new BufferedWriter(fw);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *bw*. The resource is referenced at lines 314, 316, 317, 318, 319, 329, 333, 337, 339, 341. The resource is closed at line 344. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).